### PR TITLE
#169258638 Multi-city trip request

### PR DESCRIPTION
**Pull request by hezronkimutai**
#### What does this PR do?
This PR adds a multi-city trip request
#### Description of Task to be completed?
- [x] Create a route for multi-city trip request
- [X] Create a route for posting locations
- [X] Create a route for approving the trip requests
- [X] Configured table associations
#### How should this be manually tested?
`git clone https://github.com/andela/team-odd-bn-backend`
`git checkout ft-multi-city-169258638`
`npm test`
`npm run dev-start`
- Open postman and test the end points

#### What are the relevant pivotal tracker stories?
[#169258638](https://www.pivotaltracker.com/n/projects/2407428/stories/169258638)
#### Screenshots (if appropriate)
Screen Shot 2019-11-21 at 15 21 57

Screen Shot 2019-11-21 at 15 23 13

Screen Shot 2019-11-20 at 10 05 31
Screen Shot 2019-11-20 at 10 06 41

#### Questions:

### CODE REVIEW

File: src/middlewares/verifyUser.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+import customize from '../helpers/Customize';
+
+dotenv.config();
+
+const verifyUser = (req, res, next) => {
```
Comment (NiyoEric) [inline src/middlewares/verifyUser.js]: @hezronkimutai I think you should name this function to `verifyToken`. it is good to be specific.

Comment (hezronkimutai) [reply src/middlewares/verifyUser.js]: Okey Eric, actually its like we don't need that  file because we already have Authenticate token.. let me delete it, thanks.

File: src/controllers/TripController.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * User can be able to request one-way trip
+   * @static
+   * @param {object} req request object
+   * @param {object} res response object
+   * @memberof UserController
+   * @returns {object} data
+   */
+  static async approveTrip(req, res) {
```
Comment (ivymwende) [inline src/controllers/TripController.js:58]: @hezronkimutai I think this should probably be a helper function because all trip types require approval

File: src/database/models/users.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ module.exports = (sequelize, DataTypes) => {
   }, {});
   users.associate = function(models) {
     // associations can be defined here
+    users.hasMany(models.tripRequest, {foreignKey: 'id'});
   };
   return users;
-};
+};
```
Comment (ivymwende) [inline src/database/models/users.js]: @hezronkimutai eol

File: src/database/models/triptype.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  const tripType = sequelize.define('tripType', {
+    tripType: DataTypes.STRING
+  }, {});
+  tripType.associate = function(models) {
+    // associations can be defined here
+    tripType.hasMany(models.tripRequest, {foreignKey: 'id'});
+
+  };
+  return tripType;
+};
```
Comment (ivymwende) [inline src/database/models/triptype.js]: @hezronkimutai eol

File: src/database/models/tripRequestCities.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    tripRequestId: DataTypes.INTEGER,
+    cityId: DataTypes.INTEGER
+  }, {});
+  tripRequestCities.associate = function(models) {
+    // associations can be defined here
+    tripRequestCities.belongsTo(models.tripRequest, {foreignKey: 'tripRequestId'});
+    tripRequestCities.belongsTo(models.cities, {foreignKey: 'cityId'});
+  };
+  return tripRequestCities;
+};
```
Comment (ivymwende) [inline src/database/models/tripRequestCities.js]: @hezronkimutai eol

File: src/database/models/tripRequest.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    returnDate: DataTypes.STRING
+  }, {});
+  tripRequest.associate = function(models) {
+    // associations can be defined here
+    tripRequest.belongsTo(models.users, {foreignKey: 'userId'});
+    tripRequest.belongsTo(models.tripType, {foreignKey: 'typeTrip'});
+    tripRequest.hasMany(models.tripRequestCities, {foreignKey: 'id'});
+  };
+  return tripRequest;
+};
```
Comment (ivymwende) [inline src/database/models/tripRequest.js]: @hezronkimutai eol

File: src/database/models/cities.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const cities = sequelize.define('cities', {
+    city: DataTypes.STRING
+  }, {});
+  cities.associate = function(models) {
+    cities.hasMany(models.tripRequestCities, {foreignKey: 'id'});
+  };
+  return cities;
+};
```
Comment (ivymwende) [inline src/database/models/cities.js]: @hezronkimutai eol

File: src/database/migrations/20191119162259-create-trip-request-cities.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('tripRequestCities');
+  }
+};
```
Comment (ivymwende) [inline src/database/migrations/20191119162259-create-trip-request-cities.js]: @hezronkimutai eol

File: src/database/migrations/20191119162252-create-trip-request.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('tripRequests');
+  }
+};
```
Comment (ivymwende) [inline src/database/migrations/20191119162252-create-trip-request.js]: @hezronkimutai eol

File: src/database/migrations/20191119162228-create-trip-type.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('tripTypes');
+  }
+};
```
Comment (ivymwende) [inline src/database/migrations/20191119162228-create-trip-type.js]: @hezronkimutai eol

File: src/database/migrations/20191119162236-create-cities.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE
+      }
+    });
+  },
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.dropTable('cities');
+  }
+};
```
Comment (ivymwende) [inline src/database/migrations/20191119162236-create-cities.js]: @hezronkimutai eol

File: src/database/migrations/20191119162259-create-trip-request-cities.js (hunk 8 above)
Comment (ivymwende) [inline src/database/migrations/20191119162259-create-trip-request-cities.js]: @hezronkimutai eol

File: src/database/models/cities.js (hunk 7 above)
Comment (ivymwende) [inline src/database/models/cities.js]: @hezronkimutai eol

File: src/database/models/tripRequestCities.js (hunk 5 above)
Comment (ivymwende) [inline src/database/models/tripRequestCities.js]: @hezronkimutai eol

File: src/controllers/LocationController.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import dotenv from 'dotenv';
+import customize from '../helpers/Customize';
+import { cities } from '../database/models';
+
+dotenv.config();
+/**
+ * @export
+ * @class TripController
```
Comment (NiyoEric) [inline src/controllers/LocationController.js]: It should be `@class LocationController` as opposed to write `@class TripController`.

File: src/controllers/LocationController.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+      const { name } = req.body;
+
+      const newLocation = await cities.create({
+        city: name
+      });
+
+
+      return customize.successMessage(req, res, 'Location posted successfully', newLocation, 201);
+    } catch (err) {
+      return res.status(400).json({ error: err.message });
```
Comment (NiyoEric) [inline src/controllers/LocationController.js]: Catch method should return 500 status code.

File: src/tests/tripTest.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    .send(mockData.users1)
+    .end((err, res) => {
+      Token.token = res.body.data;
+    });
+  chai.request(app)
+    .post('/api/v1/auth/signup')
+    .send(mockData.users3)
+    .end((err, res) => {
+      Token.token1 = res.body.data;
+    });
```
Comment (ivymwende) [inline src/tests/tripTest.js]: @hezronkimutai I think you should add more tests

File: src/middlewares/CheckMultiCityTrip.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+export default class CheckMultiCityTrip {
+/**
+   * validate one-way trip
+   * @static
+   * @param {object} req response Object
+   * @param {object} res request Object
+   * @param {object} next next
+   * @memberof ValidateOneWayTrip
+   * @returns {object} data
+   */
```
Comment (NiyoEric) [inline src/middlewares/CheckMultiCityTrip.js]: Update your JSDoc by mentioning the correct class and functions names used.

**Comment by ivymwende** [comment]
@hezronkimutai Also your PR title should follow the correct format as well as your PR description

File: src/routes/api/authRoute.js [hunk 16]
```diff
@@ -148,7 +152,7 @@ authRouter.post('/forgot-password', Middlewares.forgotPasswordRules(), checkForg
  *
  */
 
-authRouter.get('/:id/resend-email', resendEmailController);
+authRouter.get('/:id/resend-email', (req, res, next) => DataEngine.findOne(req, res, next, users, { id: req.params.id }), resendEmailController);
```
Comment (NiyoEric) [inline src/routes/api/authRoute.js]: you have created `DataEngine` class to save queries in one file but Line **116** and Line **155**  are making our middlewares to be complex. I advise `DataEnigne` to have helper functions that do not depend on middlewares.

File: src/routes/api/locationRoute.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+ * /locations:
+ *    post:
+ *      summary: User can request for one-way trip
+ *      tags: [Users]
+ *      requestBody:
+ *        required: true
+ *        content:
+ *          application/json:
+ *            schema:
+ *              $ref: '#/components/schemas/User'
```
Comment (NiyoEric) [inline src/routes/api/locationRoute.js]: Change the schema reference name, it should not be `User` may be `Location`.

File: src/routes/api/tripRoute.js [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+ * /trip/multi-city:
+ *    post:
+ *      summary: User can request for one-way trip
+ *      tags: [Users]
+ *      requestBody:
+ *        required: true
+ *        content:
+ *          application/json:
+ *            schema:
+ *              $ref: '#/components/schemas/User'
```
Comment (NiyoEric) [inline src/routes/api/tripRoute.js]: Change schema reference name, we already have `User` schema.

**Review by NiyoEric — CHANGES_REQUESTED** [review]
Well, job done @hezronkimutai!  I left comment to help you improve your work.

**Review by ivymwende — COMMENTED** [review]
Great Job @hezronkimutai!
However have you thought about multicity trips being in two ways?
- we can have multicity one way trip in that the user will be making multiple stops but does not have a return date
- we can have a mutlicity return trip in that the user will have multiple stops but has a return date

File: src/tests/tripTest.js (hunk 14 above)
Comment (hezronkimutai) [reply src/tests/tripTest.js]: @ivymwende, may be you should be specific on the edge case I should capture in the tests.

File: src/controllers/TripController.js (hunk 2 above)
Comment (hezronkimutai) [reply src/controllers/TripController.js:58]: Just as the sign up controller I thought this should be an approve trip controller, the only think you and @william000000  will do is to send an email containing a url to this the approval route, I dont think you need to create another approval route.

File: src/controllers/TripController.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+ * @export
+ * @class TripController
+ */
+class TripController {
+  /**
+   * User can be able to request one-way trip
+   * @static
+   * @param {object} req request object
+   * @param {object} res response object
+   * @memberof UserController
```
Comment (william000000) [inline src/controllers/TripController.js]: Rename this to `TripController`

File: src/controllers/TripController.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  static async requestTrip(req, res) {
+    try {
+      const userId = req.user.id;
+      const {
+        reason, city, startDate, returnDate
+      } = req.body;
+      const locationArr = city.split(',');
+      const newTrip = await tripRequest.create({
+        userId, reason, status: 'pending', startDate, returnDate, typeTrip: 1
```
Comment (william000000) [inline src/controllers/TripController.js]: I think you should set this status(`pending`) by default in the migration file of this table `tripRequest`;
for example:    
status: {
        type: Sequelize.STRING,
        defaultValue: 'pending'
      }

File: src/controllers/UserController.js [hunk 21]
```diff
@@ -81,11 +77,8 @@ class UserController {
      * @returns {Object} response
      */
   static async resendEmailController(req, res) {
-    const user = await users.findOne({ where: { id: req.params.id } });
-    if (!user) {
-      return Customize.errorMessage(req, res, 'Invalid ID', 400);
-    }
-    emailHelper.verifyEmailHelper(req, user.dataValues);
+    // DataEngine.findOne(req, res, users, { id: req.params.id });
```
Comment (william000000) [inline src/controllers/UserController.js]: Remove this comment

File: src/controllers/TripController.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+      return Customize.errorMessage(req, res, err.message, 500);
+    }
+  }
+
+  /**
+   * User can be able to request one-way trip
+   * @static
+   * @param {object} req request object
+   * @param {object} res response object
+   * @memberof UserController
```
Comment (william000000) [inline src/controllers/TripController.js:55]: Also, rename this as the name of its class is

File: src/database/seeders/20191119054216-users.js [hunk 23]
```diff
@@ -5,7 +5,9 @@ module.exports = {
     return queryInterface.bulkInsert('users', [{
         firstName: 'John',
         lastName: 'Doe',
-        email: 'demo@demo.com',
+        email: 'demo@gmail.com',
+        isVerified:'false',
+        signupType:'normal',
```
Comment (victkarangwa) [inline src/database/seeders/20191119054216-users.js]: @hezronkimutai  because isVerified is of Boolean type, you don't need to put a single quote.

Comment (victkarangwa) [reply src/database/seeders/20191119054216-users.js]: signupType should be 'Barefoot' instead of Normal, just the sake of being specific.

File: src/middlewares/DataEngine.js [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+   * @param {object} table table
+   * @param {object} condition bject whose key is the column and value is the value
+   * @memberof UserController
+   * @returns {object} either an error or calls next
+   */
+  static async findConflict(req, res, table, condition) {
+    const oneRow = await table.findOne({ where: condition });
+    if (oneRow) {
+      return Customize.errorMessage(req, res, 'Conflict', 409);
+    }
```
Comment (victkarangwa) [inline src/middlewares/DataEngine.js]: I think you can specify the conflict. For instance, by passing the conflict message in `find conflict` parameter

**Review by victkarangwa — COMMENTED** [review]
Nice work @hezronkimutai

File: src/controllers/TripController.js [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+   * @memberof TripController
+   * @returns {object} data
+   */
+  static async requestTrip(req, res) {
+    try {
+      const userId = req.user.id;
+      const {
+        reason, city, startDate, returnDate
+      } = req.body;
+      const locationArr = city.split(',');
```
Comment (julietezekwe) [inline src/controllers/TripController.js]: Please how are you sending the cities that needs this spliting. You should receive it as an array or cities.

**Review by julietezekwe — CHANGES_REQUESTED** [review]
Please review the comment and change the multiple cities' request body to accept an array of objects for each city request. Nice job so far

File: src/database/seeders/20191119054216-users.js (hunk 23 above)
Comment (hezronkimutai) [reply src/database/seeders/20191119054216-users.js]: Yeah, exactly, thanks Victor.

File: src/controllers/TripController.js (hunk 25 above)
Comment (hezronkimutai) [reply src/controllers/TripController.js]: Thanks for the feedback, I have implemented it.

**Review by victkarangwa — APPROVED** [review]
I liked the way you implemented the feedback @hezronkimutai 
You are good to go!

**Review by julietezekwe — APPROVED** [review]

File: src/controllers/LocationController.js [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+      });
+
+
+      return Customize.successMessage(req, res, 'Location posted successfully', newLocation, 201);
+    } catch (err) {
+      return Customize.errorMessage(req, res, err.message, 500);
+    }
+  }
+}
+export default LocationController;
```
Comment (william000000) [inline src/controllers/LocationController.js:34]: @hezronkimutai  Thanks a lot to create this for us to create  a new city without the need for seeding data

**Review by NiyoEric — APPROVED** [review]
